server/client.go

```diff
@@ -2797,6 +2800,36 @@ func isReservedReply(reply []byte) bool {
 	return false
 }
 
+// Header constants for new msg header support.
+const (
+	hdrMarker    = "⚡NATS⚡"
+	hdrMagic     = "⚡NATS⚡/0.1\r\n"
+	hdrTerm      = "\r\n\r\n"
+	hdrMarkerLen = len(hdrMarker)
+	hdrMagicLen  = len(hdrMagic)
+	minHdrLen    = len(hdrMagic) + len(hdrTerm)
+)
+
+func (c *client) checkForHeader(msg []byte) error {
+	ml := len(msg)
+	// I think these are ok, possible collision for header but no big deal IMO.
+	if ml <= hdrMarkerLen {
+		return nil
+	}
+	if ml < hdrMagicLen || !bytes.Equal(msg[:hdrMagicLen], []byte(hdrMagic)) {
```

Comment:
Payload that starts with byte 226 (and keep in mind that payload could be binary data where byte 226 is not at all unreasonable) will be rejected with bad header, as long as the payload is at least 9 bytes long (not counting the \r\n of the MSG protocol that is included in len(msg).
Modified nats-pub to simply replace msg[0] with value 226, and otherwise use the rest of value from the command line parameter. Trying to send "123456789" (again replacing first byte with 226) causes the publish to fail.

Reply:
Yes understood this was a collision avoidance tactic for real life use.

Reply:
AFAIK the proposed solution you and Matthias have may help but I think there are degenerate cases there as well that may be worse TBH.

Reply:
Our protocol is text base, but not the user payload. Having a serie of bytes is not far fetched. It could be protobuf (like Streaming is using) or some byte payload. We can't assume that it will never start with 226. Unless I misunderstand what you are saying..

Reply:
I did a bit of research to look at what I thought were common payloads. So JSON protobuf, binary assets like png, jpg, etc and pdfs. None of these start with 226 nor would they match with the full header unless on purpose or some malicious behavior. 

That being said I understand the point you both are trying to raise and am thinking through it as you are.

Reply:
I am simply saying that we should not fail if it is not a "valid" header in this approach. If it does not start with the magic header, it is simply not a header, that's all. What we should be doing with this approach is try to find the end of the header so we can skip for connections without support for it.
In other words, the condition i have commented should return nil instead of BadHeader, meaning it is not a header but simple payload.

Reply:
We do find the end, agree that is main goal for server and then having a sense if client supports it or not.

I think for client development if it really looks like a header but does not have proper term we should error. We can discuss why more this afternoon.

Reply:
We have decided on proto changes for clients (and possibly server to server).